source/test/TestApp/PlayFabEventTest.cpp

```diff
@@ -210,14 +211,67 @@ namespace PlayFabUnit
 
     void PlayFabEventTest::PrivateMemberCallbackTest(TestContext& testContext)
     {
-        eventTestContext = std::make_shared<TestContext*>(&testContext);
+       eventTestContext = std::make_shared<TestContext*>(&testContext);
 
-        std::shared_ptr<PlayFabEventAPI*> api = SetupEventTest();
+        eventApi = SetupEventTest();
 
-        (*api)->EmitEvent(MakeEvent(0, PlayFabEventType::Default),
+        (*eventApi)->EmitEvent(MakeEvent(0, PlayFabEventType::Default),
         std::bind(&PlayFabEventTest::NonStaticEmitEventCallback, this, std::placeholders::_1, std::placeholders::_2));
     }
 
+    void PlayFabEventTest::GenericMultiThreadedTest(uint32_t pNumThreads, uint32_t pNumEventsPerThread)
+    {
+        std::atomic<uint32_t> eventsRemaining = pNumThreads * pNumEventsPerThread;
+        for (uint32_t thread = 0; thread < pNumThreads; ++thread)
+        {
+            testThreadPool.emplace_back(
+                [&eventsRemaining, pNumEventsPerThread, this]() 
+                {
+                    std::shared_ptr<PlayFabEventAPI*> api = SetupEventTest();
+                    for (uint32_t i = 0; i < pNumEventsPerThread; ++i)
+                    {
+                        (*api)->EmitEvent(MakeEvent(0, PlayFabEventType::Default),
+                            [&eventsRemaining, pNumEventsPerThread]
+                            (std::shared_ptr<const PlayFab::IPlayFabEvent>, std::shared_ptr<const PlayFab::IPlayFabEmitEventResponse>)
+                            {
+                                eventsRemaining--;
+                                if (eventsRemaining == 0)
```

Comment:
>eventsRemaining [](start = 36, length = 15)

I think what you had before was correct.
You want to do an atomic write-then-read

by splitting the `--` from the 0 check it's no longer atomic. Multiple separate threads might trip the if-check now

Reply:
this was a bad merge from a different branch. Thanks for catching this :)

---
In reply to: [346620630](https://github.com/PlayFab/XPlatCoreTemplate/pull/44#discussion_r346620630) [](ancestors = 346620630)